### PR TITLE
Minimal effort  GAT prediction in docking workflow. 

### PR DESCRIPTION
**Pull request by hmacdope**
Adds a very basic attempt at GAT prediction in the current docking workflow.  Thought I would do an MVP first and separate from an in-progress a rewrite of the docking script.

### CODE REVIEW

**Comment by hmacdope** [comment]
This MVP of adding the GAT prediction to the docking workflow is probably ready for a review @kaminow @JenkeScheen

File: asapdiscovery-ml/asapdiscovery/ml/inference/inference.py [hunk 1]
```diff
@@ -173,8 +176,39 @@ def predict(self, g: dgl.DGLGraph):
...
+        """
+        if isinstance(smiles, str):
+            smiles = [smiles]
+
+        gids = GraphInferenceDataset(
+            smiles, node_featurizer=CanonicalAtomFeaturizer(), **kwargs
+        )
+
+        data = [self.predict(g) for g in gids]
+        return np.concatenate(np.asarray(data))
```
Comment (JenkeScheen) [inline asapdiscovery-ml/asapdiscovery/ml/inference/inference.py:214]: in the case of a single SMILES, does the predicted value get returned as `value` or as `[value]`? I think returning as a float would make more sense?

File: asapdiscovery-ml/asapdiscovery/ml/tests/test_inference.py [hunk 2]
```diff
@@ -107,3 +106,22 @@ def test_gatinference_predict_dataset_equal(
...
+    g1_infds, g2_infds, gids = test_inference_data
+    # same data different smiles order
+    assert inference_cls is not None
+    smiles = list(gids.smiles_dict.keys())
+    assert len(smiles) == 2
+    s1 = smiles[0]
+    s2 = smiles[1]
+    output1 = inference_cls.predict(gids[s1])
+    output2 = inference_cls.predict(gids[s2])
+    assert_allclose(output1, output2)
```
Comment (JenkeScheen) [inline asapdiscovery-ml/asapdiscovery/ml/tests/test_inference.py:127]: Are we testing that predicted values are different from each other? Sometimes with poorly trained models they end up learning to predict the mean label (because they're not able to learn some form of inference so predicting the mean minimizes the cost fn) of the training set and will consistently predict that value.

**Review by JenkeScheen — COMMENTED** [review]
This is great! For project support I'll probably rename the property to "ML_SCORE" or something (just so that chemists are less confused) but that would be too granular for this code, I think. 

I'll let @kaminow have a look at the `GraphInferenceDataset(Dataset)` class.

File: asapdiscovery-ml/asapdiscovery/ml/tests/test_inference.py (hunk 2 above)
Comment (hmacdope) [reply asapdiscovery-ml/asapdiscovery/ml/tests/test_inference.py:127]: I can add a test for two distinct smiles and make sure they are different. I think thats a good idea.

File: asapdiscovery-ml/asapdiscovery/ml/inference/inference.py (hunk 1 above)
Comment (hmacdope) [reply asapdiscovery-ml/asapdiscovery/ml/inference/inference.py:214]: Yeah its being returned as [value], I can change to returning `value` for a single smiles.

File: asapdiscovery-docking/asapdiscovery/docking/scripts/run_docking_oe.py [hunk 3]
```diff
@@ -124,7 +127,7 @@ def load_dus(file_base, by_compound=False):
     return dataset_dict, du_dict
 
 
-def mp_func(out_dir, lig_name, du_name, *args, **kwargs):
+def mp_func(out_dir, lig_name, du_name, GAT_model, *args, **kwargs):
```
Comment (kaminow) [inline asapdiscovery-docking/asapdiscovery/docking/scripts/run_docking_oe.py:130]:
```suggestion
def mp_func(out_dir, lig_name, du_name, *args, GAT_model=None, **kwargs):
```
not sure if this is actually necessary, but just to make it clear that the `GAT_model` isn't necessary

File: asapdiscovery-docking/asapdiscovery/docking/scripts/run_docking_oe.py [hunk 4]
```diff
@@ -137,6 +140,8 @@ def mp_func(out_dir, lig_name, du_name, *args, **kwargs):
         Ligand name
     du_name : str
         DesignUnit name
+    GAT_model : GATInference
```
Comment (kaminow) [inline asapdiscovery-docking/asapdiscovery/docking/scripts/run_docking_oe.py:143]:
```suggestion
    GAT_model : GATInference, optional
```
same was with above

File: asapdiscovery-docking/asapdiscovery/docking/scripts/run_docking_oe.py [hunk 5]
```diff
@@ -333,6 +342,10 @@ def main():
         raise ValueError("Need to specify exactly one of --exp_file or --lig_file.")
     n_mols = len(mols)
 
+    # load ml models
+
+    GAT_model = GATInference("model1")
```
Comment (kaminow) [inline asapdiscovery-docking/asapdiscovery/docking/scripts/run_docking_oe.py:347]: I think we should probably make running the prediction optional, eg with some boolean CLI flag or something

File: asapdiscovery-ml/asapdiscovery/ml/dataset.py [hunk 6]
```diff
@@ -547,22 +564,35 @@ def __init__(
...
+            return self.smiles_dict[idx]
+        elif isinstance(idx, int):
+            return self.graphs[idx]
+        elif isinstance(idx, list):
+            return [self.graphs[i] for i in idx]
+        elif isinstance(idx, slice):
+            # Use the slice indices to slice the list of graphs and create a new GraphInferenceDataset
+            start, stop, step = idx.indices(len(self.graphs))
+            indices = range(start, stop, step)
+            new_dataset = GraphInferenceDataset(
```
Comment (kaminow) [inline asapdiscovery-ml/asapdiscovery/ml/dataset.py:586]: I'm not sure which is better here, but `pytorch` has a [`Subset` class](https://pytorch.org/docs/stable/data.html#torch.utils.data.Subset) that I think does what we'd want in the case of getting a `slice` here, and it also prevents any data copying I believe

File: asapdiscovery-ml/asapdiscovery/ml/dataset.py [hunk 7]
```diff
@@ -547,22 +564,35 @@ def __init__(
...
+                exp_compounds=[],  # empty list since we're creating a new dataset
```
Comment (kaminow) [inline asapdiscovery-ml/asapdiscovery/ml/dataset.py:587]: why don't we want to copy over the compounds?

File: asapdiscovery-ml/asapdiscovery/ml/inference/inference.py [hunk 8]
```diff
@@ -173,8 +176,43 @@ def predict(self, g: dgl.DGLGraph):
...
             DGLGraph object.
 
+        Returns
+        -------
+        np.ndarray
+            Predictions for each graph.
         """
         with torch.no_grad():
             output_tensor = self.model(g, g.ndata["h"])
+            # unsure exactly why this reshape is done
```
Comment (kaminow) [inline asapdiscovery-ml/asapdiscovery/ml/inference/inference.py:186]: this was done in the training methods to match the shape of the labels, probably isn't needed here

File: asapdiscovery-ml/asapdiscovery/ml/inference/inference.py [hunk 9]
```diff
@@ -173,8 +176,43 @@ def predict(self, g: dgl.DGLGraph):
...
+        """
+        if isinstance(smiles, str):
+            smiles = [smiles]
+
+        gids = GraphInferenceDataset(
+            smiles, node_featurizer=CanonicalAtomFeaturizer(), **kwargs
+        )
+
+        data = [self.predict(g) for g in gids]
+        data = np.concatenate(np.asarray(data))
```
Comment (kaminow) [inline asapdiscovery-ml/asapdiscovery/ml/inference/inference.py:214]: is there a reason we need the `np.concatenate` here? I don't think it matters, I'm just curious

File: asapdiscovery-ml/asapdiscovery/ml/tests/data/make_example_GAT_data.py [hunk 10]
```diff
@@ -9,13 +9,15 @@
 
 # lets make a synthetic dataset using a smiles from Fragalysis
 
-# we will use a smiles from P0045_0A:TRY-UNI-2EDDB1FF-7 Fragalysis ID
+# we will use a smiles from P0045_0A:TRY-UNI-2EDDB1FF-7 Fragalysis ID for compound 1 and reordered version
 input_smiles = "Cc1ccncc1NC(=O)Cc1cc(Cl)cc(O[C@H]2CC(=O)N2)c1"
 
 mol = Chem.MolFromSmiles(input_smiles)
 # randomly shuffle the smiles string to test equivariance
 reorder_smiles = Chem.MolToSmiles(mol, doRandom=True)
```
Comment (kaminow) [inline asapdiscovery-ml/asapdiscovery/ml/tests/data/make_example_GAT_data.py:17]: this is a great test!

**Review by kaminow — CHANGES_REQUESTED** [review]
a couple of questions/comments but overall this looks great!

File: asapdiscovery-ml/asapdiscovery/ml/inference/inference.py (hunk 9 above)
Comment (hmacdope) [reply asapdiscovery-ml/asapdiscovery/ml/inference/inference.py:214]: good spotting, the `asarray` probably does the job.

Comment (hmacdope) [reply asapdiscovery-ml/asapdiscovery/ml/inference/inference.py:214]: Actually wrong, the concatenate is needed to flatten.

File: asapdiscovery-ml/asapdiscovery/ml/dataset.py (hunk 7 above)
Comment (hmacdope) [reply asapdiscovery-ml/asapdiscovery/ml/dataset.py]: Fixed now by using subset.

File: asapdiscovery-ml/asapdiscovery/ml/dataset.py (hunk 6 above)
Comment (hmacdope) [reply asapdiscovery-ml/asapdiscovery/ml/dataset.py]: Using subset now, good tip seems very handy.

**Comment by hmacdope** [comment]
@kaminow hopefully I have addressed your review. :)
